docs/weave-tutorial/05-messages.md

```diff
@@ -0,0 +1,109 @@
+
+```go
+type Tx interface {
+    Persistent
+    // GetMsg returns the action we wish to communicate
+    GetMsg() (Msg, error)
+}
+```
+
+And every application can extend it with additional functionality, such as [Signatures](https://godoc.org/github.com/iov-one/weave/x/sigs#SignedTx), [Fees](https://godoc.org/github.com/iov-one/weave/x/cash#FeeTx), or anything else your application needs. The data placed in the Transaction is meant to be anything that applies to all modules, and is processed by a **Middleware**.
+
+A [Message](https://godoc.org/github.com/iov-one/weave#Msg) is also persistent and can be pretty much anything that an extension defines, as it also defines the Handler to process it. The only necessary feature of a Message is that it can return a `Path() string` which allows us to route it to the proper Handler.
+
+When we define a concrete transaction type for one application, we define it in protobuf with a set of possible messages that it can contain. Every application can add optional field to the transaction and allow a different set of messages, and the Handlers and Decorators work orthogonally to this, regardless of the **concrete** Transaction type.
+
+## Defining Messages
+
+Messages are similar to the `POST` endpoints in a typical API. They are the only way to effect a change in the system. Ignoring the issue of authentication and rate limitation, which is handled by the Decorators / Middleware, when we design Messages, we focus on all possible state transitions and the information they need to proceed.
+
+In the Orderbook example, we can imagine:
+
+- Create market
+- Create order book
+- Create order
+
+## Dive into Code
+
+First create your `msg.go` file. This is where the magic will happen.
+
+First create an __init()__ function and register to migration schema:
+
+```go
+func init() {
+    // Migration needs to be registered for every message introduced in the codec.
+    migration.MustRegister(1, &CreateOrderBookMsg{}, migration.NoModification)
+}
+```
+
+Define path that will be used for routing messages to handler:
+
+```go
+func (CreateOrderBookMsg) Path() string {
+    return "orderbook/create_orderbook"
+}
+```
+
+After that ensure your message is a `weave.Msg`:
+
+```go
+var _ weave.Msg = (*CreateOrderBookMsg)(nil)
+```
+
+## Validation
+
+While validation for data models is much more like SQL constraints: “max length 20”, “not null”, “constaint foo > 3”, validation for messages is validating potentially malicious data coming in from external sources and should be validated more thoroughly. One may want to use regexp to avoid control characters or null bytes in a “string” input. Maybe restrict it to alphanumeric or ascii characters, strip out html, or allow full utf-8. Addresses must be checked to be the valid length. Amount being sent to be positive (else I send you -5 ETH and we have a **TakeMsg**, instead of **SendMsg**).
```

Comment:
Maybe it is worth mentioning that `Validate` method on a message must only provide a sanity check for the data it represents and must not rely on any external state. Message can only ensure the data format and hardcoded logic. It cannot validate business logic. Business logic is validated in a handler.